### PR TITLE
4 controllers support + osContCh optimization

### PR DESCRIPTION
**Pull request by Arceveti**
- Barebones 4 controllers support
- osContCh to skip polling ports after the last used port (no skipping yet)
- MAX_NUM_PLAYERS define to limit the number of controllers (must be at least 2 until the gamecube controller fixes)

### CODE REVIEW

**Review by thecozies — APPROVED** [review]

File: src/game/game_init.c [hunk 1]
```diff
@@ -88,10 +85,11 @@ u16 sRenderingFramebuffer = 0;
 // Goddard Vblank Function Caller
 void (*gGoddardVblankCallback)(void) = NULL;
 
-// Defined controller slots
-struct Controller *gPlayer1Controller = &gControllers[0];
-struct Controller *gPlayer2Controller = &gControllers[1];
-struct Controller *gPlayer3Controller = &gControllers[2]; // Probably debug only, see note below
+// Defined controller slots. Anything above MAX_NUM_PLAYERS will be unused.
+struct Controller* const gPlayer1Controller = &gControllers[0];
```
Comment (thecozies) [inline src/game/game_init.c:89]: since player one can change to pointing at any controller, I'd switch these back to not be const

**Review by thecozies — CHANGES_REQUESTED** [review]
small change, partial review

**Review by gheskett — APPROVED** [review]
LGTM I think, not gonna bother with rigorous testing on this one if the other PR is still planned anyway.

File: src/game/game_init.c [hunk 2]
```diff
@@ -677,34 +664,43 @@ void init_controllers(void) {
...
+        if (cont >= MAX_NUM_PLAYERS) {
+            break;
+        }
+
         // Is controller plugged in?
         if (gControllerBits & (1 << port)) {
             // The game allows you to have just 1 controller plugged
             // into any port in order to play the game. this was probably
             // so if any of the ports didn't work, you can have controllers
             // plugged into any of them and it will work.
```
Comment (thecozies) [inline src/game/game_init.c:678]: unfortunately this functionality isnt working for me, i have port 1 completely unplugged and a controller in port 2

**Review by thecozies — CHANGES_REQUESTED** [review]
small regression

**Comment by gheskett** [comment]
Moving to 2.2/3.0.

File: src/game/game_init.c (hunk 2 above)
Comment (Arceveti) [reply src/game/game_init.c:678]: Fixed, empty ports weren't advancing the PIF read pointer in osContGetReadDataEx.

**Review by thecozies — APPROVED** [review]

**Comment by gheskett** [comment]
> Moving to 2.2/3.0.

nvm

**Pull request merged**